File: app/src/main/java/com/miaadrajabi/menuxmlparser/java/domain/usecase/MenuEditor.java
package com.miaadrajabi.menuxmlparser.java.domain.usecase;

import com.miaadrajabi.menuxmlparser.java.domain.model.MenuModels.*;

import java.util.ArrayList;
import java.util.List;

public class MenuEditor {
    public DynamicMenus addGroup(DynamicMenus menus, MenuGroup group) {
        List<MenuGroup> groups = new ArrayList<>(menus.groups);
        groups.add(group);
        return new DynamicMenus(groups);
    }

    public DynamicMenus removeGroup(DynamicMenus menus, String groupId) {
        List<MenuGroup> groups = new ArrayList<>();
        for (MenuGroup g : menus.groups) if (!g.id.equals(groupId)) groups.add(g);
        return new DynamicMenus(groups);
    }

    public DynamicMenus updateGroup(DynamicMenus menus, MenuGroup updated) {
        List<MenuGroup> groups = new ArrayList<>();
        for (MenuGroup g : menus.groups) groups.add(g.id.equals(updated.id) ? updated : g);
        return new DynamicMenus(groups);
    }

    public DynamicMenus addTopMenu(DynamicMenus menus, String groupId, TopLevelMenu top) {
        List<MenuGroup> groups = new ArrayList<>();
        for (MenuGroup g : menus.groups) {
            if (g.id.equals(groupId)) {
                List<TopLevelMenu> tops = new ArrayList<>(g.topLevelMenus);
                tops.add(top);
                groups.add(copyGroup(g, tops));
            } else groups.add(g);
        }
        return new DynamicMenus(groups);
    }

    public DynamicMenus updateTopMenu(DynamicMenus menus, String groupId, TopLevelMenu top) {
        List<MenuGroup> groups = new ArrayList<>();
        for (MenuGroup g : menus.groups) {
            if (g.id.equals(groupId)) {
                List<TopLevelMenu> tops = new ArrayList<>();
                for (TopLevelMenu t : g.topLevelMenus) tops.add(t.id.equals(top.id) ? top : t);
                groups.add(copyGroup(g, tops));
            } else groups.add(g);
        }
        return new DynamicMenus(groups);
    }

    public DynamicMenus removeTopMenu(DynamicMenus menus, String groupId, String topId) {
        List<MenuGroup> groups = new ArrayList<>();
        for (MenuGroup g : menus.groups) {
            if (g.id.equals(groupId)) {
                List<TopLevelMenu> tops = new ArrayList<>();
                for (TopLevelMenu t : g.topLevelMenus) if (!t.id.equals(topId)) tops.add(t);
                groups.add(copyGroup(g, tops));
            } else groups.add(g);
        }
        return new DynamicMenus(groups);
    }

    public DynamicMenus addOperator(DynamicMenus menus, String topId, MenuItem operator) {
        return mapTop(menus, topId, t -> {
            List<MenuItem> ops = new ArrayList<>(t.items);
            ops.add(operator);
            return copyTop(t, ops);
        });
    }

    public DynamicMenus updateOperator(DynamicMenus menus, String topId, MenuItem operator) {
        return mapTop(menus, topId, t -> {
            List<MenuItem> ops = new ArrayList<>();
            for (MenuItem m : t.items) ops.add(m.id.equals(operator.id) ? operator : m);
            return copyTop(t, ops);
        });
    }

    public DynamicMenus removeOperator(DynamicMenus menus, String topId, String operatorId) {
        return mapTop(menus, topId, t -> {
            List<MenuItem> ops = new ArrayList<>();
            for (MenuItem m : t.items) if (!m.id.equals(operatorId)) ops.add(m);
            return copyTop(t, ops);
        });
    }

    private interface TopMapper { TopLevelMenu apply(TopLevelMenu t); }

    private DynamicMenus mapTop(DynamicMenus menus, String topId, TopMapper mapper) {
        List<MenuGroup> groups = new ArrayList<>();
        for (MenuGroup g : menus.groups) {
            List<TopLevelMenu> tops = new ArrayList<>();
            boolean changed = false;
            for (TopLevelMenu t : g.topLevelMenus) {
                if (t.id.equals(topId)) {
                    tops.add(mapper.apply(t));
                    changed = true;
                } else tops.add(t);
            }
            groups.add(changed ? copyGroup(g, tops) : g);
        }
        return new DynamicMenus(groups);
    }

    private MenuGroup copyGroup(MenuGroup g, List<TopLevelMenu> tops) {
        return new MenuGroup(g.id, g.englishTitle, g.persianTitle, g.isEnabled, g.icon, tops);
    }

    private TopLevelMenu copyTop(TopLevelMenu t, List<MenuItem> items) {
        return new TopLevelMenu(t.id, t.englishTitle, t.persianTitle, t.icon, t.isEnabled, t.isExpandable, items, t.services);
    }
}

File: app/src/main/java/com/miaadrajabi/menuxmlparser/domain/usecase/MenuEditor.kt
package com.miaadrajabi.menuxmlparser.domain.usecase

import com.miaadrajabi.menuxmlparser.domain.model.*

class MenuEditor {
    fun addGroup(menus: DynamicMenus, group: MenuGroup): DynamicMenus =
        menus.copy(groups = menus.groups + group)

    fun removeGroup(menus: DynamicMenus, groupId: String): DynamicMenus =
        menus.copy(groups = menus.groups.filterNot { it.id == groupId })

    fun updateGroup(menus: DynamicMenus, updated: MenuGroup): DynamicMenus =
        menus.copy(groups = menus.groups.map { if (it.id == updated.id) updated else it })

    fun addTopMenu(menus: DynamicMenus, groupId: String, top: TopLevelMenu): DynamicMenus =
        menus.copy(groups = menus.groups.map { g ->
            if (g.id == groupId) g.copy(topLevelMenus = g.topLevelMenus + top) else g
        })

    fun updateTopMenu(menus: DynamicMenus, groupId: String, top: TopLevelMenu): DynamicMenus =
        menus.copy(groups = menus.groups.map { g ->
            if (g.id == groupId) g.copy(topLevelMenus = g.topLevelMenus.map { if (it.id == top.id) top else it }) else g
        })

    fun removeTopMenu(menus: DynamicMenus, groupId: String, topId: String): DynamicMenus =
        menus.copy(groups = menus.groups.map { g ->
            if (g.id == groupId) g.copy(topLevelMenus = g.topLevelMenus.filterNot { it.id == topId }) else g
        })

    fun addOperator(menus: DynamicMenus, topId: String, operator: MenuItem): DynamicMenus =
        menus.copy(groups = menus.groups.map { g ->
            g.copy(topLevelMenus = g.topLevelMenus.map { t ->
                if (t.id == topId) t.copy(items = t.items + operator) else t
            })
        })

    fun updateOperator(menus: DynamicMenus, topId: String, operator: MenuItem): DynamicMenus =
        menus.copy(groups = menus.groups.map { g ->
            g.copy(topLevelMenus = g.topLevelMenus.map { t ->
                if (t.id == topId) t.copy(items = t.items.map { if (it.id == operator.id) operator else it }) else t
            })
        })

    fun removeOperator(menus: DynamicMenus, topId: String, operatorId: String): DynamicMenus =
        menus.copy(groups = menus.groups.map { g ->
            g.copy(topLevelMenus = g.topLevelMenus.map { t ->
                if (t.id == topId) t.copy(items = t.items.filterNot { it.id == operatorId }) else t
            })
        })
}


